### PR TITLE
fix: add JSON body API key extraction for @librechat/agents ≥3.1.74 compatibility

### PR DESCRIPTION
**Pull request by gafda**
### Problem
On LibreChat v0.8.5, `@librechat/agents` 3.1.74 silently removed the `X-API-Key` header from outbound requests to KubeCodeRun, spreading params into the JSON body instead. This causes **401 Unauthorized** on every `execute_code` tool invocation.

### Solution
This PR introduces the following changes:

1. **API Key Extraction**:
   - Adds body-based API key extraction as a fallback in both `AuthenticationMiddleware` and `SecurityMiddleware`. The key resolution priority is:
     - `X-API-Key` header (existing, unchanged — agents ≤3.1.73)
     - `Authorization: Bearer` header (existing, unchanged)
     - JSON body fields: `LIBRECHAT_CODE_API_KEY`, `api_key`, `apiKey` (new)
   - Body extraction only triggers for `POST`/`PUT`/`PATCH` when no header key is found. A buffer-and-replay pattern ensures the request body remains available for downstream route handlers.

2. **Trusted Networks Bypass**:
   - Adds a feature to bypass authentication for trusted networks within the cluster, improving compatibility for in-cluster callers.

3. **Error Handling and Logging**:
   - Improves error handling and logging during warm pod creation and replenishment, making debugging easier.

4. **Request Body Size Limit**:
   - Enforces a limit on the request body size to prevent potential abuse or resource exhaustion.

### Testing
- 37 new unit tests covering all extraction paths, body buffering, replay semantics, priority ordering, and edge cases.
- All existing auth/security middleware tests pass (zero regressions).
- Lint (`ruff check`) and format (`ruff format --check`) pass cleanly.
- No conflicts with `feat-pod-labels-config` branch (verified via merge dry-run).

### Security
- No API key values logged.
- Existing constant-time comparison via `AuthenticationService.validate_api_key` unchanged.
- Body parsing is read-only; does not consume the stream for downstream handlers.

### CODE REVIEW

File: src/middleware/auth.py [hunk 1]
```diff
@@ -108,6 +118,55 @@ def _extract_api_key(self, request: Request) -> str | None:
...
+        body_parts: list[bytes] = []
+        while True:
+            message = await receive()
+            body = message.get("body", b"")
+            if body:
+                body_parts.append(body)
+            if not message.get("more_body", False):
+                break
+
+        full_body = b"".join(body_parts)
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/middleware/auth.py:209]: Added `_MAX_AUTH_BODY_BYTES` (1 MB) cap — raises **413** when exceeded.

File: src/middleware/auth.py [hunk 2]
```diff
@@ -40,7 +41,15 @@ async def __call__(self, scope: dict, receive: Callable, send: Callable):
...
 
         try:
-            await self._authenticate_request(request, scope)
+            # Try header-based extraction first
+            api_key = self._extract_api_key(request)
+
+            # If no key in headers, try JSON body extraction for POST/PUT/PATCH
+            if api_key is None and request.method in ("POST", "PUT", "PATCH"):
+                body_bytes, receive = await self._buffer_body(receive)
+                api_key = self._extract_api_key_from_body(body_bytes)
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/middleware/auth.py:58]: Added **Content-Type guard** — only buffers when `application/json`.

File: src/middleware/security.py [hunk 3]
```diff
@@ -232,6 +242,55 @@ def _extract_api_key(self, request: Request) -> str | None:
...
+        body_parts: list[bytes] = []
+        while True:
+            message = await receive()
+            body = message.get("body", b"")
+            if body:
+                body_parts.append(body)
+            if not message.get("more_body", False):
+                break
+
+        full_body = b"".join(body_parts)
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/middleware/security.py:336]: Same 1 MB cap with **413** rejection.

File: src/middleware/security.py [hunk 4]
```diff
@@ -94,7 +95,15 @@ async def send_wrapper(message):
 
             # Handle authentication (skip for excluded paths and OPTIONS)
             if not self._should_skip_auth(request):
-                await self._authenticate_request(request, scope)
+                # Try header-based extraction first
+                api_key = self._extract_api_key(request)
+
+                # If no key in headers, try JSON body extraction for POST/PUT/PATCH
+                if api_key is None and request.method in ("POST", "PUT", "PATCH"):
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/middleware/security.py]: Same Content-Type guard — skips multipart/form-data, text/plain, etc.

**Comment by pedrojreis** [comment]
Hey, any idea when this will be merged?! KubeCodeRun doesnt work on Librechat 0.8.5 :)

File: src/middleware/auth.py [hunk 5]
```diff
@@ -64,12 +81,47 @@ async def __call__(self, scope: dict, receive: Callable, send: Callable):
 
     def _should_skip_auth(self, request: Request) -> bool:
         """Check if authentication should be skipped."""
-        return request.url.path in self.excluded_paths or request.method == "OPTIONS"
+        if request.url.path in self.excluded_paths or request.method == "OPTIONS":
+            return True
+
+        # Bypass auth for requests from trusted networks (e.g. in-cluster callers)
+        if self._trusted_networks and self._is_trusted_network(request):
+            return True
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/middleware/auth.py:89]: Now uses `request.client.host` (socket peer) for trust decisions.

File: src/middleware/security.py [hunk 6]
```diff
@@ -139,17 +159,52 @@ async def _validate_request(self, request: Request):
...
             or path.startswith("/api/v1/admin")
             or path.startswith("/admin-dashboard")
             or request.method == "OPTIONS"
-        )
+        ):
+            return True
+
+        # Bypass auth for requests from trusted networks (e.g. in-cluster callers)
+        if self._trusted_networks and self._is_trusted_network(request):
+            return True
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/middleware/security.py:172]: Now uses `request.client.host` (socket peer) for trust decisions.

File: src/services/kubernetes/pool.py [hunk 7]
```diff
@@ -155,12 +156,25 @@ async def _warmup(self):
...
             tasks = [self._create_warm_pod() for _ in range(min(batch_size, needed - i))]
-            await asyncio.gather(*tasks, return_exceptions=True)
+            results = await asyncio.gather(*tasks, return_exceptions=True)
+            for result in results:
+                if isinstance(result, BaseException):
+                    logger.error(
+                        "Unexpected error during pool warmup",
+                        language=self.language,
+                        error=str(result),
+                        exc_info=result,
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/services/kubernetes/pool.py]: Changed to `exc_info=(type(result), result, result.__traceback__)`

File: src/services/kubernetes/pool.py [hunk 8]
```diff
@@ -342,7 +373,15 @@ async def _replenish_loop(self):
...
                         tasks = [self._create_warm_pod() for _ in range(batch)]
-                        await asyncio.gather(*tasks, return_exceptions=True)
+                        results = await asyncio.gather(*tasks, return_exceptions=True)
+                        for result in results:
+                            if isinstance(result, BaseException):
+                                logger.error(
+                                    "Unexpected error during pool replenishment",
+                                    language=self.language,
+                                    error=str(result),
+                                    exc_info=result,
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/services/kubernetes/pool.py]: Changed to `exc_info=(type(result), result, result.__traceback__)`

File: tests/unit/test_trusted_networks.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    }
+    request = Request(scope)
+    request._url = MagicMock()
+    request._url.path = path
+    # Mock client
+    request._client = MagicMock()
+    request._client.host = client_ip
+    # Patch scope client for Request.client property
+    scope["client"] = (client_ip, 12345)
+    return Request(scope)
```
[the remark this answers is not in the document]
Comment (gafda) [reply tests/unit/test_trusted_networks.py]: Now returns the original `request` with `scope["client"]` set properly.

File: tests/unit/test_auth_body_extraction.py [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+def auth_middleware(mock_app):
+    """Create auth middleware instance."""
+    return AuthenticationMiddleware(mock_app)
+
+
+@pytest.fixture
+def security_middleware(mock_app):
+    """Create security middleware instance."""
+    with patch("src.middleware.security.settings") as mock_settings:
+        mock_settings.max_file_size_mb = 10
```
[the remark this answers is not in the document]
Comment (gafda) [reply tests/unit/test_auth_body_extraction.py:35]: Added `mock_settings.auth_trusted_networks = ""`

File: src/middleware/auth.py [hunk 11]
```diff
@@ -64,12 +81,47 @@ async def __call__(self, scope: dict, receive: Callable, send: Callable):
...
+    async def _authenticate_request(self, request: Request, scope: dict, *, api_key: str | None = None):
         """Handle API key authentication."""
-        # Extract API key
-        api_key = self._extract_api_key(request)
+        # Use provided api_key or extract from headers as fallback
+        if api_key is None:
+            api_key = self._extract_api_key(request)
 
         # Get authentication service
         auth_service = await get_auth_service()
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/middleware/auth.py:134]: Added `if not api_key: raise HTTPException(401, "Missing API key")` in both middlewares

File: src/middleware/security.py [hunk 12]
```diff
@@ -94,7 +101,20 @@ async def send_wrapper(message):
...
+                # If no key in headers, try JSON body extraction for POST/PUT/PATCH
+                content_type = request.headers.get("content-type", "").split(";", 1)[0].strip().lower()
+                if (
+                    api_key is None
+                    and request.method in ("POST", "PUT", "PATCH")
+                    and content_type == "application/json"
+                ):
+                    body_bytes, receive = await self._buffer_body(receive)
+                    api_key = self._extract_api_key_from_body(body_bytes)
+
```
[the remark this answers is not in the document]
Comment (gafda) [reply src/middleware/security.py:116]: Added `if not api_key: raise HTTPException(401, "Missing API key")` in both middlewares

**Review by aron-muon — APPROVED** [review]
